### PR TITLE
Fix #57354: Fix-it for deprecated initializers removes the `.init` part from `self.init`

### LINKED ISSUE #57354: [SR-15026] Fix-it for deprecated initializers removes the `.init` part from `self.init`
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-15026      |
   |Radar             | rdar://81572131         |
   |Original Reporter | @WowbaggersLiquidLunch      |
   |Type              | Bug    |

Attachment: [Download](https://user-images.githubusercontent.com/2727770/164964277-a592ab30-b14f-4e94-bce5-4271e9569087.gz)

      
<details>
  <summary>Environment</summary>

Swift 5.5 (swiftlang-1300.0.27.6 clang-1300.0.27.2)  
Xcode 13.0 beta 4 (13A5201i)  
macOS 12.0 Beta (21A5294g)

</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | Bug, DiagnosticsQoI, StarterBug        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: e74bf2f692b142ab23f91cd5f9b6813d

  </details>

**Issue Description:**

This happened when I was replacing the call sites of a deprecated initializer in SwiftPM on top of the commit 0a17db09d7065be8a8ad925a4cfa50f3dd306cfc. I'm not sure if it's an Xcode problem or a diagnostics problem, so I'm filing for the bug both here and on the Feedback Assistant (FB9457368).

Please see the attached screen recording.

### PR DESCRIPTION
**Pull request by cbjeukendrup**
Now, the fix-it will only remove the `.init` part if the expr at the left side of the dot is a type, and not, for example, a `self` expr.

Resolves: #57354

### CODE REVIEW

File: test/attr/attr_availability.swift [hunk 1]
```diff
@@ -590,6 +590,17 @@ func testFactoryMethods() {
...
+  @available(*, deprecated, renamed: "init(new:)")
+  init(old: Int) {}
+
+  init(new: Int) {}
+
+  init(test: Int) {
+    self.init(old: test) // expected-warning {{'init(old:)' is deprecated: replaced by 'init(new:)'}} expected-note {{use 'init(new:)' instead}} {{15-18=new}}
+  }
+}
+
```
Comment (LucianoPAlmeida) [inline test/attr/attr_availability.swift:603]: Should we add a test case for `super` as well? 
```swift
class A {
  @available(*, deprecated, renamed: "init(new:)")
  init(old: Int) {}
  
  init(new: Int) {}
  
  init(test: Int) {
  }
}
class B: A {
  override init(test: Int) {
     super.init(old: test) // expected-warning {{'init(old:)' is deprecated: replaced by 'init(new:)'}} expected-note {{use 'init(new:)' instead}} {{15-18=new}}
   }
}
```
Not sure this `isa<TypeExpr>` logic going to catch this ...

Comment (cbjeukendrup) [reply test/attr/attr_availability.swift:605]: Good idea, I added one! 

I also noticed the following: obviously, when subclasses inherit initializers from their superclass, the availability attribute on that initializer is inherited too. But it turned out that the message and rename information is _not_ inherited by the generated availability attribute. So I fixed that in a second commit.

**Comment by hborla** [comment]
@cbjeukendrup thank you for your contribution! @xymus and @tshortli know the most about the availability checker; I've tagged them to help review this change

**Comment by tshortli** [comment]
@swift-ci please test

**Comment by tshortli** [comment]
@swift-ci please test macOS

**Review by tshortli — APPROVED** [review]
Looks good to me, thank you for this contribution!

**Comment by cbjeukendrup** [comment]
Thanks for reviewing, @tshortli!

**Pull request merged**